File: 1_two_sum.kt
// https://leetcode.com/problems/two-sum/description/

// O(n^2) solution, because we need to traverse twice
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        for (i in nums.indices) {
            val requiredPair = target - nums[i]
            
            for (j in (i + 1) until nums.size) {
                if (nums[j] == requiredPair) return intArrayOf(i, j)
            }
        }

        throw IllegalStateException("Invalid state, shouldn't have been reachable.")
    }
}

// O(n) solution. :) 
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val traversalMap = mutableMapOf<Int, Int>() // Key: Number; Value: Index

        for (i in nums.indices) {
            if (traversalMap[target - nums[i]] != null) {
                return intArrayOf(i, traversalMap[target - nums[i]]!!) // If memory complexity wasn't an issue this wouldn't be here as a separate variable can be tested outside of the if statement and the compiler would know this is a non-null at this point.
            } else {
                traversalMap.put(nums[i], i)
            }
        }

        throw IllegalStateException("Invalid state, shouldn't have been reachable.")
    }
}

File: 1_two_sum.cpp
// https://leetcode.com/problems/two-sum/

// Quick and dirty O(n^2) solution
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); i++) {
            int currentNum = nums[i];
            int targetNum = target - currentNum;
            
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[j] == targetNum) {
                    return {i, j};
                }
            }
        }
        
        throw invalid_argument("Function was supplied with invalid inputs as there always must be exactly one correct solution.");
    }
};


// O(n) solution. Lesson: It's hashmaps. It always is.
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> traversalMap; // Map: <value, index>

        for (int i = 0; i < nums.size(); i++) {
            int complementerValue = target - nums[i]; // We are looking for this value

            if (traversalMap.find(complementerValue) != traversalMap.end()) { // Lookup if map has key of COMPLEMENTER
                return {i, traversalMap[complementerValue]};
            } else {
                traversalMap[nums[i]] = i;
            }
        }
        
        throw invalid_argument("Invalid state, shouldn't have been reachable.");
    }
};
